src/clojure/nrepl/server.clj

```diff
@@ -133,11 +135,64 @@
   java.io.Closeable
   (close [this] (stop-server this)))
 
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind ^Integer port] (InetSocketAddress. bind port))
+        ;; We fallback to 127.0.0.1 instead of to localhost to avoid
+        ;; a dependency on the order of ipv4 and ipv6 records for
+        ;; localhost in /etc/hosts
+        bind (or bind "127.0.0.1")]
+    (doto (ServerSocket.)
+      (.setReuseAddress true)
+      (.bind (addr bind port)))))
+
+
+;;; Unix domain (file namespace) sockets
+
+(defn- class-finder [full-path]
+  (memoize
+   (fn []
+     (try
+       (import full-path)
+       (catch ClassNotFoundException ex
+         nil)))))
+
+(def unix-socket-address-class
+  (class-finder 'org.newsclub.net.unix.AFUNIXSocketAddress))
+
+(def unix-server-socket-class
+  (class-finder 'org.newsclub.net.unix.AFUNIXServerSocket))
+
+(defn fs-socket
+  "Returns a filesystem socket bound to the path if the dependencies (which are
+  loaded dynamically) are available.  If
+  com.kohlschutter.junixsocket/junixsocket-core is not available, throws the
+  ex-info map {:nrepl/kind ::no-filesystem-sockets}."
+  [path]
+  (let [address-class (unix-socket-address-class)
+        socket-class (unix-server-socket-class)]
+    (when-not (and address-class socket-class)
+      (let [msg "Listening on filesystem sockets requires junixsocket dependency"]
+        (log msg)
+        (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets}))))
+    (let [addr (clojure.lang.Reflector/invokeConstructor
+                address-class
+                (into-array java.io.File [(java.io.File. path)]))
+          new-instance (.getDeclaredMethod socket-class "newInstance" nil)
+          new-instance #(.invoke new-instance nil nil)]
+      (doto (new-instance)
+        (.bind addr)))))
+
+
 (defn start-server
   "Starts a socket-based nREPL server.  Configuration options include:
 
    * :port — defaults to 0, which autoselects an open port
    * :bind — bind address, by default \"127.0.0.1\"
+   * :socket - filesystem socket path (alternative to :port and :bind)
```

Comment:
I'm guessing that here we can recycle `:bind` as well and just check if it's a path. Too many decisions. 😆

Reply:
We could, if you prefer -- be happy to rework all this to just rely on some kind of "/foo" detection if you like, and regarding the extra lines, I suspect one of those is a mistake, but otherwise, I'd used two blank lines to separate the unix domain stuff.  Happy to remove that, though.

Reply:
Oh, and I may nearly have JDK 16 sockets working too -- or rather, I think I do have them working (though we may have to go back to `import*` (not sure yet)), but I'm having a little trouble getting `lein` to handle `lein test` for (opportunistic) JDK socket testing) and `lein with-profiles test-junixsocket test` for junixsocket.  The profile merge is causing some unexpected errors.  I'll get back to you once I investigate further, and of course, we can always add that later, once I sort it out.

Reply:
Got it, I think.  Should have an update in the next day or two.

Reply:
> We could, if you prefer -- be happy to rework all this to just rely on some kind of "/foo" detection if you like, and regarding the extra lines, I suspect one of those is a mistake, but otherwise, I'd used two blank lines to separate the unix domain stuff. Happy to remove that, though.

Btw, should we always pass the socket explicitly? I've been thinking that for the server it also makes some sense to just create a socket file automatically if one is not specified. I seem to recall some services operated in this manner. Perhaps I'm obsessing over the details too much, but given how hard it is to change something in nREPL once it's released it's better to be safe than sorry. :D

Reply:
Could you elaborate a bit - not sure I understand what you're considering here yet.

Reply:
And no worries at all about the details, I completely understand for something like this, and am happy to try to help settle on something we think might be durable.